### PR TITLE
Disabling submit button until form is touched

### LINKED ISSUE #13243: Disable Create/Update Buttons Until Form Changes Are Made Across All Forms
**Is your feature request related to a problem? Please describe.**
Yes. Currently, the **Create** and **Update** buttons remain enabled by default in many forms, even when no changes have been made or no input has been provided. This may result in unnecessary API calls or seeing misleading success toasts.

This feature has already been implemented in some parts of the app, so this request is to ensure consistent behavior across all forms.

**Pages where this should be implemented:**
1. **Add Device**
- Create form only (update already handles this)
2. **Specimen Definitions**
- Create and Update
3. **Observation Definitions**
- Create and Update
4. **Activity Definitions**
- Create and Update
5. **Billing**
- Create Discount Code
- Create Discount Component
6. **Charge Item Definitions**
- Create and Update
7. **Healthcare Services**
- Create and Update
8. **Product Knowledge**
- Create form only
9. **Product**
- Create form only

**Describe the solution you'd like**
The Create or Update button should be disabled by default until valid changes are made in the Form.

**Describe alternatives you've considered**

**Additional context**

https://github.com/user-attachments/assets/1a579377-1b98-4273-b8a2-d4392b3931ae

---
### Instructions for Requesting Assignment:

To request assignment, please clearly outline your solution and timeline by commenting on the issue using the format below:

**Describe your solution clearly:**
Provide a detailed explanation of your proposed solution, including your approach, key implementation steps, and relevant examples or references. Mention any dependencies, assumptions, or risks you foresee that might affect your timeline or implementation.

**Expected Timeline:**
- End date: [Expected submission date of a completed Pull Request]

**Additional Context:**
Include any other relevant context, links, screenshots, or resources that support your proposed solution.

> 🚨 Your assignment may be unassigned if there is no activity or progress within the stated timeline unless communicated clearly and agreed upon.

### PR DESCRIPTION
**Pull request by GSMPRANEETH**
## Proposed Changes

Fixes #13243 
- Disabling button using `!form.formState.isDirty`
- Modified defaultValues to have all possible fields have default values
- The component `RoleForm.tsx` has been refactored and this mechanism is implemented in PR #13851 
- Refactoring `PlugConfigEdit.tsx`, now uses `useForm` hook, made `isDirty` check for submit button, using Form Messages, formSchema and zodResolver
- The components `SpecimenDefinitionForm.tsx` & `ProductKnowledgeForm.tsx`, using form.reset() in useEffect

image
image
image
Screenshot from 2025-11-26 16-32-15
image

Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [x] Add specs that demonstrate the bug or test the new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is placed in I18n files.
- [x] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [x] Request peer reviews.
- [x] Complete QA on mobile devices.
- [x] Complete QA on desktop devices.
- [x] Add or update Playwright tests for related changes

## Summary by CodeRabbit

* **Bug Fixes**
  * Submit buttons across many forms now disable unless the form is dirty or a submission is pending, preventing accidental submissions of unchanged data.
  * Form initialization flows consolidated for more consistent create/edit defaults.

* **New Features**
  * Added optional metadata and additional device fields to capture more device details.

* **Tests**
  * Updated automated tests to reflect stricter dirty-check and required-field flows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx [hunk 1]
```diff
@@ -217,14 +217,17 @@ function ProductKnowledgeFormContent({
...
+      slug_value: "",
       product_type: ProductKnowledgeType.medication,
+      alternate_identifier: "",
       names: [],
       storage_guidelines: [],
-      code: null,
+      code: undefined,
       base_unit: undefined,
-      definitional: null,
+      definitional: { intended_routes: [] },
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx]: fixed

File: src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx [hunk 2]
```diff
@@ -320,6 +322,10 @@ function ProductKnowledgeFormContent({
     }
   }
 
+  React.useEffect(() => {
+    console.log("dirty fields: ", form.formState.dirtyFields);
+  });
+
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx]: removed

File: tests/facility/settings/tokenCategory/tokenCategoryCreate.spec.ts [hunk 3]
```diff
@@ -115,7 +115,8 @@ test.describe("Token Category Create - Permission Tests", () => {
         /\/facility\/[^/]+\/settings\/token_category\/new/,
       );
 
-      // Click Create button without filling any fields
+      // Click Create button after making for dirty but invalid
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply tests/facility/settings/tokenCategory/tokenCategoryCreate.spec.ts]: fixed

File: src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx [hunk 4]
```diff
@@ -151,6 +151,25 @@ export function SpecimenDefinitionForm({
...
+      patient_preparation: initialData?.patient_preparation ?? [],
+      collection: initialData?.collection ?? undefined,
+      type_tested: initialData?.type_tested ?? {
+        is_derived: false,
+        preference: Preference.preferred,
+        container: {},
+        single_use: false,
+      },
+    });
+  }, [form, initialData]);
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx]: fixed

File: src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx [hunk 5]
```diff
@@ -235,6 +238,10 @@ function ProductKnowledgeFormContent({
     reValidateMode: "onChange",
   });
 
+  React.useEffect(() => {
+    form.reset(() => getDefaultValues());
+  }, [form, existingData]);
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx]: fixed

File: src/pages/Facility/settings/product/ProductForm.tsx [hunk 6]
```diff
@@ -508,7 +508,10 @@ export function ProductFormContent({
             <Button type="button" variant="outline" onClick={onCancel}>
               {t("cancel")}
             </Button>
-            <Button type="submit" disabled={isPending}>
+            <Button
+              type="submit"
+              disabled={isPending || !form.formState.isDirty}
+            >
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/product/ProductForm.tsx:519]: no problem

File: src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx [hunk 7]
```diff
@@ -151,6 +151,25 @@ export function SpecimenDefinitionForm({
...
+      description: initialData?.description ?? "",
+      derived_from_uri: initialData?.derived_from_uri ?? undefined,
+      type_collected: initialData?.type_collected ?? undefined,
+      patient_preparation: initialData?.patient_preparation ?? [],
+      collection: initialData?.collection ?? undefined,
+      type_tested: initialData?.type_tested ?? {
+        is_derived: false,
+        preference: Preference.preferred,
+        container: {},
+        single_use: false,
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx]: works as intended

File: src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx [hunk 8]
```diff
@@ -236,11 +235,8 @@ function ProductKnowledgeFormContent({
   });
 
   React.useEffect(() => {
-    if (isEditMode && existingData) {
-      form.reset(getDefaultValues());
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [existingData?.slug, isEditMode]);
+    form.reset(getDefaultValues());
+  }, [existingData]);
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/productKnowledge/ProductKnowledgeForm.tsx:240]: no

File: src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx [hunk 9]
```diff
@@ -151,12 +151,15 @@ export function SpecimenDefinitionForm({
...
 
+  // s
+
   const { fields, append, remove, update } = useFieldArray({
     control: form.control,
     name: "patient_preparation",
   });
 
   React.useEffect(() => {
+    form.reset();
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx:160]: works as intended

File: src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx [hunk 10]
```diff
@@ -151,12 +151,15 @@ export function SpecimenDefinitionForm({
     },
   });
 
+  // s
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/specimen-definitions/SpecimenDefinitionForm.tsx]: will be removed

File: src/pages/Facility/settings/product/ProductForm.tsx [hunk 11]
```diff
@@ -229,6 +229,11 @@ export function ProductFormContent({
         : {
             status: ProductStatusOptions.active,
             product_knowledge: slug,
+            charge_item_definition: undefined,
+            batch: {
+              lot_number: "",
+            },
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Facility/settings/product/ProductForm.tsx:235]: k

**Comment by nihal467** [comment]
removing th testing label since the PR is draft

**Comment by GSMPRANEETH** [comment]
@nihal467 @Jacobjeevan @rithviknishad 
The components `RoleForm.tsx` & `PlugConfigEdit.tsx` doesn't use useForm hook, so no check implemented
for these components, should i implement the useForm hook so they follow similar implementation of forms as others?
Also with this i can implement the intended functionality of disabling the submit button?

File: tests/facility/services/locations/inventory/toDispatch.spec.ts [hunk 12]
```diff
@@ -94,6 +94,9 @@ test.describe("Facility To-Dispatch Orders Inventory Flow", () => {
     await expect(tableRow1).toContainText("Paracetamol");
     await expect(tableRow1).toContainText("5");
     await page.getByRole("link", { name: "Create Delivery Order" }).click();
+    await page
+      .getByRole("textbox", { name: "Note" })
+      .fill("To make the form dirty");
```
Comment (nihal467) [inline tests/facility/services/locations/inventory/toDispatch.spec.ts]: @GSMPRANEETH look into the coderabbit comments

File: tests/facility/settings/chargeItemDefinition/chargeItemDefinitionsCreate.spec.ts [hunk 13]
```diff
@@ -40,6 +40,9 @@ test.describe("Charge Item Definition Creation", () => {
 
   test("validate required fields", async ({ page }) => {
     await page.getByRole("button", { name: /add definition/i }).click();
+    await page
+      .getByRole("textbox", { name: "description" })
+      .fill("To make the form dirty");
```
Comment (nihal467) [inline tests/facility/settings/chargeItemDefinition/chargeItemDefinitionsCreate.spec.ts]: use faker

**Review by nihal467 — COMMENTED** [review]
- dont use hard coded text, you faker

**Comment by nihal467** [comment]
image

- the error message of category in the observtion is broken
- dont add default selected value for data type
- LOINC error message is missing

File: src/pages/Apps/PlugConfigEdit.tsx [hunk 14]
```diff
@@ -60,10 +80,9 @@ export function PlugConfigEdit({ slug }: Props) {
...
 
-  const handleSubmit = async (e: React.FormEvent) => {
-    e.preventDefault();
-    const meta = JSON.parse(config.meta);
-    const configPayload = { ...config, meta };
+  const handleSubmit = async (data: Props) => {
+    const meta = data.meta ? JSON.parse(data.meta) : {};
+    const configPayload = { ...data, meta };
     upsertConfig(configPayload);
   };
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Apps/PlugConfigEdit.tsx]: using suggested change

File: src/pages/Apps/PlugConfigEdit.tsx [hunk 15]
```diff
@@ -60,10 +80,9 @@ export function PlugConfigEdit({ slug }: Props) {
     onSuccess: () => navigate("/admin/apps"),
   });
 
-  const handleSubmit = async (e: React.FormEvent) => {
-    e.preventDefault();
-    const meta = JSON.parse(config.meta);
-    const configPayload = { ...config, meta };
+  const handleSubmit = async (data: Props) => {
+    const meta = data.meta ? JSON.parse(data.meta) : {};
```
[the remark this answers is not in the document]
Comment (GSMPRANEETH) [reply src/pages/Apps/PlugConfigEdit.tsx]: using suggestion

File: tests/facility/settings/chargeItemDefinition/chargeItemDefinitionsCreate.spec.ts (hunk 13 above)
Comment (GSMPRANEETH) [reply tests/facility/settings/chargeItemDefinition/chargeItemDefinitionsCreate.spec.ts]: implemented faker

**Review by nihal467 — CHANGES_REQUESTED** [review]
-  You should be adding new test to verify that when the form is empty, the button is disabled, and upon any field is filled, the button is enabled
- Fix the playwright test failing on your PR
